File: backend/internal/service/claude_web_safety.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
)

const ClaudeWebSafetyPresetVersion = 1

type ClaudeWebSafetyPreset struct {
	BaseRPM                 int
	BaseTPM                 int
	MaxSessions             int
	WindowCostLimit         float64
	WindowCostStickyReserve float64
}

var claudeWebSafetyPresets = map[string]ClaudeWebSafetyPreset{
	ClaudeTierPro:    {BaseRPM: 6, BaseTPM: 80000, MaxSessions: 2, WindowCostLimit: 8, WindowCostStickyReserve: 1},
	ClaudeTierMax5x:  {BaseRPM: 12, BaseTPM: 160000, MaxSessions: 4, WindowCostLimit: 40, WindowCostStickyReserve: 5},
	ClaudeTierMax20x: {BaseRPM: 20, BaseTPM: 240000, MaxSessions: 6, WindowCostLimit: 80, WindowCostStickyReserve: 10},
}

var claudeWebManagedUsageKeys = map[string]struct{}{
	"quota_used": {}, "quota_daily_used": {}, "quota_weekly_used": {},
	"quota_daily_start": {}, "quota_weekly_start": {},
	"quota_daily_reset_at": {}, "quota_weekly_reset_at": {},
}

func ClaudeWebSafetyPresetForTier(tier string) (ClaudeWebSafetyPreset, bool) {
	preset, ok := claudeWebSafetyPresets[NormalizeClaudeTier(tier)]
	return preset, ok
}

func NormalizeClaudeWebSafetyExtra(extra map[string]any, applyDefaults bool) (map[string]any, error) {
	normalized := make(map[string]any, len(extra)+10)
	for key, value := range extra {
		if _, managed := claudeWebManagedUsageKeys[key]; !managed {
			normalized[key] = value
		}
	}

	tier := ClaudeTierPro
	tierExplicit := false
	if raw, exists := normalized["claude_tier"]; exists {
		tierExplicit = true
		value, ok := raw.(string)
		if !ok || NormalizeClaudeTier(value) == "" {
			return nil, errors.New("invalid Claude tier")
		}
		tier = NormalizeClaudeTier(value)
	}
	normalized["claude_tier"] = tier

	if raw, exists := normalized["claude_tier_source"]; exists {
		source, ok := raw.(string)
		if !ok || !isClaudeTierSource(source) {
			return nil, errors.New("invalid Claude tier source")
		}
		normalized["claude_tier_source"] = strings.TrimSpace(source)
	} else {
		if tierExplicit {
			normalized["claude_tier_source"] = ClaudeTierSourceManual
		} else {
			normalized["claude_tier_source"] = ClaudeTierSourceInferred
		}
	}

	positiveKeys := []string{
		"base_rpm", "base_tpm", "max_sessions", "session_idle_timeout_minutes",
		"window_cost_limit", "window_cost_sticky_reserve", "quota_limit",
		"quota_daily_limit", "quota_weekly_limit",
	}
	for _, key := range positiveKeys {
		if raw, exists := normalized[key]; exists {
			value, ok := claudeWebPositiveNumber(raw)
			if !ok {
				return nil, fmt.Errorf("%s must be a positive finite number", key)
			}
			normalized[key] = value
		}
	}
	for _, key := range []string{"rpm_strategy", "tpm_strategy"} {
		if raw, exists := normalized[key]; exists {
			value, ok := raw.(string)
			if !ok || (value != "tiered" && value != "sticky_exempt") {
				return nil, fmt.Errorf("invalid %s", key)
			}
		}
	}

	if applyDefaults {
		preset := claudeWebSafetyPresets[tier]
		setClaudeWebDefault(normalized, "base_rpm", preset.BaseRPM)
		setClaudeWebDefault(normalized, "base_tpm", preset.BaseTPM)
		setClaudeWebDefault(normalized, "max_sessions", preset.MaxSessions)
		setClaudeWebDefault(normalized, "session_idle_timeout_minutes", 5)
		setClaudeWebDefault(normalized, "window_cost_limit", preset.WindowCostLimit)
		setClaudeWebDefault(normalized, "window_cost_sticky_reserve", preset.WindowCostStickyReserve)
		setClaudeWebDefault(normalized, "rpm_strategy", "tiered")
		setClaudeWebDefault(normalized, "tpm_strategy", "tiered")
		normalized["claude_safety_preset_version"] = ClaudeWebSafetyPresetVersion
	}
	return normalized, nil
}

func setClaudeWebDefault(extra map[string]any, key string, value any) {
	if _, exists := extra[key]; !exists {
		extra[key] = value
	}
}

func isClaudeTierSource(value string) bool {
	switch strings.TrimSpace(value) {
	case ClaudeTierSourceManual, ClaudeTierSourceProfile, ClaudeTierSourceInferred, ClaudeTierSourceProfileDefault:
		return true
	default:
		return false
	}
}

func claudeWebPositiveNumber(value any) (any, bool) {
	var number float64
	switch v := value.(type) {
	case int:
		number = float64(v)
	case int32:
		number = float64(v)
	case int64:
		number = float64(v)
	case float32:
		number = float64(v)
	case float64:
		number = v
	case json.Number:
		var err error
		number, err = v.Float64()
		if err != nil {
			return nil, false
		}
	case string:
		var err error
		number, err = strconv.ParseFloat(strings.TrimSpace(v), 64)
		if err != nil {
			return nil, false
		}
	default:
		return nil, false
	}
	if math.IsNaN(number) || math.IsInf(number, 0) || number <= 0 {
		return nil, false
	}
	if math.Trunc(number) == number {
		return int(number), true
	}
	return number, true
}

File: frontend/src/components/account/claudeWebSafety.ts
export type ClaudeTier = '' | 'Pro' | 'Max_5x' | 'Max_20x'
export type ClaudeTierSource = '' | 'manual' | 'profile' | 'inferred' | 'profile_default'
export type ClaudeLimitStrategy = 'tiered' | 'sticky_exempt'

export interface ClaudeWebSafetyState {
  tier: ClaudeTier
  tierSource: ClaudeTierSource
  windowCostEnabled: boolean
  windowCostLimit: number | null
  windowCostStickyReserve: number | null
  sessionLimitEnabled: boolean
  maxSessions: number | null
  sessionIdleTimeoutMinutes: number | null
  rpmLimitEnabled: boolean
  baseRPM: number | null
  rpmStrategy: ClaudeLimitStrategy
  tpmLimitEnabled: boolean
  baseTPM: number | null
  tpmStrategy: ClaudeLimitStrategy
  quotaLimit: number | null
  quotaDailyLimit: number | null
  quotaWeeklyLimit: number | null
}

export const CLAUDE_WEB_SAFETY_PRESET_VERSION = 1
export const CLAUDE_WEB_SAFETY_CONFIG_KEYS = [
  'claude_tier', 'claude_tier_source', 'claude_safety_preset_version',
  'window_cost_limit', 'window_cost_sticky_reserve',
  'max_sessions', 'session_idle_timeout_minutes',
  'base_rpm', 'rpm_strategy', 'base_tpm', 'tpm_strategy',
  'quota_limit', 'quota_daily_limit', 'quota_weekly_limit'
] as const

export const CLAUDE_WEB_SAFETY_PRESETS = {
  Pro: { baseRPM: 6, baseTPM: 80000, maxSessions: 2, windowCostLimit: 8, windowCostStickyReserve: 1 },
  Max_5x: { baseRPM: 12, baseTPM: 160000, maxSessions: 4, windowCostLimit: 40, windowCostStickyReserve: 5 },
  Max_20x: { baseRPM: 20, baseTPM: 240000, maxSessions: 6, windowCostLimit: 80, windowCostStickyReserve: 10 }
} as const

export function createClaudeWebSafetyState(): ClaudeWebSafetyState {
  return {
    tier: '',
    tierSource: '',
    windowCostEnabled: false,
    windowCostLimit: null,
    windowCostStickyReserve: null,
    sessionLimitEnabled: false,
    maxSessions: null,
    sessionIdleTimeoutMinutes: null,
    rpmLimitEnabled: false,
    baseRPM: null,
    rpmStrategy: 'tiered',
    tpmLimitEnabled: false,
    baseTPM: null,
    tpmStrategy: 'tiered',
    quotaLimit: null,
    quotaDailyLimit: null,
    quotaWeeklyLimit: null
  }
}

export function hydrateClaudeWebSafety(extra?: Record<string, unknown>): ClaudeWebSafetyState {
  const state = createClaudeWebSafetyState()
  if (!extra) return state
  const tier = extra.claude_tier
  if (tier === 'Pro' || tier === 'Max_5x' || tier === 'Max_20x') state.tier = tier
  const source = extra.claude_tier_source
  if (source === 'manual' || source === 'profile' || source === 'inferred' || source === 'profile_default') state.tierSource = source
  state.windowCostLimit = positiveNumber(extra.window_cost_limit)
  state.windowCostStickyReserve = positiveNumber(extra.window_cost_sticky_reserve)
  state.windowCostEnabled = state.windowCostLimit != null
  state.maxSessions = positiveNumber(extra.max_sessions)
  state.sessionIdleTimeoutMinutes = positiveNumber(extra.session_idle_timeout_minutes)
  state.sessionLimitEnabled = state.maxSessions != null
  state.baseRPM = positiveNumber(extra.base_rpm)
  state.rpmLimitEnabled = state.baseRPM != null
  state.rpmStrategy = extra.rpm_strategy === 'sticky_exempt' ? 'sticky_exempt' : 'tiered'
  state.baseTPM = positiveNumber(extra.base_tpm)
  state.tpmLimitEnabled = state.baseTPM != null
  state.tpmStrategy = extra.tpm_strategy === 'sticky_exempt' ? 'sticky_exempt' : 'tiered'
  state.quotaLimit = positiveNumber(extra.quota_limit)
  state.quotaDailyLimit = positiveNumber(extra.quota_daily_limit)
  state.quotaWeeklyLimit = positiveNumber(extra.quota_weekly_limit)
  return state
}

export function applyClaudeWebPreset(state: ClaudeWebSafetyState, tier: Exclude<ClaudeTier, ''>, options: { overwrite?: boolean } = {}): void {
  const preset = CLAUDE_WEB_SAFETY_PRESETS[tier]
  const overwrite = options.overwrite === true
  state.tier = tier
  state.tierSource = 'manual'
  state.windowCostEnabled = true
  state.sessionLimitEnabled = true
  state.rpmLimitEnabled = true
  state.tpmLimitEnabled = true
  if (overwrite || state.windowCostLimit == null) state.windowCostLimit = preset.windowCostLimit
  if (overwrite || state.windowCostStickyReserve == null) state.windowCostStickyReserve = preset.windowCostStickyReserve
  if (overwrite || state.maxSessions == null) state.maxSessions = preset.maxSessions
  if (overwrite || state.sessionIdleTimeoutMinutes == null) state.sessionIdleTimeoutMinutes = 5
  if (overwrite || state.baseRPM == null) state.baseRPM = preset.baseRPM
  if (overwrite || state.baseTPM == null) state.baseTPM = preset.baseTPM
}

export function serializeClaudeWebSafety(state: ClaudeWebSafetyState): Record<string, unknown> {
  const extra: Record<string, unknown> = {}
  if (state.tier) {
    extra.claude_tier = state.tier
    extra.claude_tier_source = state.tierSource || 'manual'
    extra.claude_safety_preset_version = CLAUDE_WEB_SAFETY_PRESET_VERSION
  }
  if (state.windowCostEnabled && positiveNumber(state.windowCostLimit) != null) {
    extra.window_cost_limit = state.windowCostLimit
    if (positiveNumber(state.windowCostStickyReserve) != null) extra.window_cost_sticky_reserve = state.windowCostStickyReserve
  }
  if (state.sessionLimitEnabled && positiveNumber(state.maxSessions) != null) {
    extra.max_sessions = state.maxSessions
    extra.session_idle_timeout_minutes = positiveNumber(state.sessionIdleTimeoutMinutes) ?? 5
  }
  if (state.rpmLimitEnabled && positiveNumber(state.baseRPM) != null) {
    extra.base_rpm = state.baseRPM
    extra.rpm_strategy = state.rpmStrategy
  }
  if (state.tpmLimitEnabled && positiveNumber(state.baseTPM) != null) {
    extra.base_tpm = state.baseTPM
    extra.tpm_strategy = state.tpmStrategy
  }
  if (positiveNumber(state.quotaLimit) != null) extra.quota_limit = state.quotaLimit
  if (positiveNumber(state.quotaDailyLimit) != null) extra.quota_daily_limit = state.quotaDailyLimit
  if (positiveNumber(state.quotaWeeklyLimit) != null) extra.quota_weekly_limit = state.quotaWeeklyLimit
  return extra
}

function positiveNumber(value: unknown): number | null {
  const parsed = typeof value === 'number' ? value : Number(value)
  return Number.isFinite(parsed) && parsed > 0 ? parsed : null
}
